image/image.go

```diff
@@ -403,6 +403,23 @@ func neededDefaultProviders(info *snap.Info) (cps []string) {
 	return cps
 }
 
+// hasBase checks if the given snap has a base in the given localInfos and
+// snaps. If not an error is returned.
+func hasBase(snap *snap.Info, local *localInfos, snaps []string) error {
+	// snap needs no base: nothing to do
+	if snap.Base == "" {
+		return nil
+	}
+	// core provides everything that core16 needs
```

Comment:
While I see the logic this will have unintended consequences.

1. boot a system with core
2. run a snap app using core as implicit effective base even though the snap demands core16
3. install core16
4. run the snap app from step 2. again

What will snap-confine be told? Will it discard the mount namespace because root filesystem has changed?
When we invoke snap-confine today, if you don't provide --base you will get the "base is core" semantics.
Note that base is *never* empty as far as snap-confine is concerned.

I may return to this comment later during the review.

Reply:
We do discard the namespace if  a base was refreshed, so I think it would maybe ok to discard it if we go from aliased core16 to core,  vs real core16.  As long as we did pivot in both cases.